src/libraries/Microsoft.Extensions.Options.DataAnnotations/src/DataAnnotationValidateOptions.cs

```diff
@@ -52,19 +55,86 @@ public ValidateOptionsResult Validate(string? name, TOptions options)
+        {
+            if (visited is not null && visited.Contains(options))
             {
-                errors.Add($"DataAnnotation validation failed for '{typeName}' members: '{string.Join(",", result.MemberNames)}' with the error: '{result.ErrorMessage}'.");
+                return true;
             }
 
-            return ValidateOptionsResult.Fail(errors);
+            results?.Clear();
+
+            bool res = Validator.TryValidateObject(options, new ValidationContext(options), results, validateAllProperties: true);
+            if (!res)
+            {
+                errors ??= new List<string>();
+
+                foreach (ValidationResult result in results!)
+                {
+                    errors.Add($"DataAnnotation validation failed for '{qualifiedName}' members: '{string.Join(",", result.MemberNames)}' with the error: '{result.ErrorMessage}'.");
+                }
+            }
+
+            foreach (PropertyInfo propertyInfo in options.GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public))
+            {
+                if (propertyInfo.GetMethod is null)
+                {
+                    continue;
+                }
+
+                object? value = propertyInfo!.GetValue(options);
+
+                if (value is null)
+                {
+                    continue;
+                }
+
+                if (propertyInfo.GetCustomAttribute<ValidateObjectMembersAttribute>() is not null)
+                {
+                    visited ??= new HashSet<object>(new ReferenceComparer());
+                    visited.Add(options);
+
+                    results ??= new List<ValidationResult>();
+                    res = TryValidateOptions(value, $"{qualifiedName}.{propertyInfo.Name}", results, ref errors, ref visited) && res;
+                }
+                else if (propertyInfo.GetCustomAttribute<ValidateEnumeratedItemsAttribute>() is not null &&
+                        value is IEnumerable enumerable)
+                {
+                    visited ??= new HashSet<object>(new ReferenceComparer());
+                    visited.Add(options);
+                    results ??= new List<ValidationResult>();
+
+                    int index = 0;
+                    foreach (object item in enumerable)
+                    {
+                        res = TryValidateOptions(item, $"{qualifiedName}.{propertyInfo.Name}[{index++}]", results, ref errors, ref visited) && res;
+                    }
+                }
+            }
+
+            return res;
+        }
+
+        private readonly struct ReferenceComparer : IEqualityComparer<object>
+        {
+            public ReferenceComparer() { }
+            bool IEqualityComparer<object>.Equals(object? x, object? y) => object.ReferenceEquals(x, y);
+            int IEqualityComparer<object>.GetHashCode(object obj) => obj is null ? 0 : obj.GetHashCode();
```

Comment:
`obj.GetHashCode()` might have been overridden by the type so you should use the `CompilerServices` helper instead. Even though this is still functionally correct for most use cases you want to avoid stacking multiple equal values (that are not reference equal) in the same bucket, and there's also the low chance that an incorrectly implemented `GetHashCode()` might break your code.
```suggestion
            int IEqualityComparer<object>.GetHashCode(object obj) => RuntimeHelpers.GetHashCode(obj);
```